docs/language-server/DocumentEventSyncPublisherSubscriberDesign.md

```diff
@@ -0,0 +1,114 @@
+# Document event sync publisher subscriber
+
+## Overview
+
+This document contains information about the implementation of a notification system inside the language server using the 
+PubSub architecture. Goal is to allow communication between language server and extensions synchronously.
+
+Following diagram shows how this has been implemented. 
+
+![alt text](../images/language-server/001-DocumentEventPubSub-class-diagram.png)
+
+- Event Subscribers and Event Publishers get loaded using JavaSPI and get subscribed to publishers when a document 
+opens or changes.
+- Event Publishers and Subscribers have an Event kind and a subscriber can only subscribe to a single publisher at once.
+If we need a subscriber to subscribe for two publishers, then we have to create two different subscribers and subscribe
+separately.
+- When an Event Publisher publishes, the `onEvent()` method is getting triggered for each subscriber that has been
+subscribed to the publisher.
+- We can override `publish()` and `onEvent()` methods to control the event synchronization.
+  
+
+## How to Use
+
+- **Event Publisher** is an internal interface and cannot be accessed publicly by extension developers. The server 
+decides which events to be delegated to a particular publisher. In the `publish()` method a debouncing mechanism has 
+implemented for the **Project Update Event Publisher**.
+- **Event Subscriber** is an interface which can be accessed publicly by extension developers and get notified when 
+subscribed to a publisher.
+- If the extension uses different file schemes, users have to handle that from extension itself.
+  (ex: expression file scheme)
+
+## Example
+
+### 1. Implementing a Subscriber
+
+We can implement `EventSubscriber` interface as follows.
+
+```java
+@JavaSPIService("org.ballerinalang.langserver.commons.eventsync.spi.EventSubscriber")
+public class CommandRegisterSubscriber implements EventSubscriber {
+    public static final String NAME = "Command register subscriber";
+
+    @Override
+    public EventKind eventKind() {
+        return EventKind.PROJECT_UPDATE;
+    }
+    
+    @Override
+    public void onEvent(ExtendedLanguageClient client, DocumentServiceContext context,
+                        LanguageServerContext languageServerContext) {
+        checkAndRegisterCommands(context);
```

Comment:
```suggestion
        // Your implementation here
```
No need to copy the same logic here. Just put a link to an implementation class (like CommandRegisteringSubscriber`.

Reply:
Also mention that the `DocumentServiceContext` and the `fileUri` can be used to determine `scheme` specific events. Currently LS supports `file` and `expr` schemes.